File: C++/NqueenProblem.cpp
#define N 4 

#include <stdbool.h> 

#include <stdio.h> 

//function to print the solution

void printSolution(int board[N][N]) 

{ 

    for (int i = 0; i < N; i++) { 

        for (int j = 0; j < N; j++) 

            printf(" %d ", board[i][j]); 

        printf("\n"); 

    } 

} 

 // function to check whether the position is safe or not 

bool isSafe(int board[N][N], int row, int col) 

{ 

    int i, j; 

    for (i = 0; i < col; i++) 

        if (board[row][i]) 

            return false; 

    for (i = row, j = col; i >= 0 && j >= 0; i--, j--) 

        if (board[i][j]) 

            return false; 

    for (i = row, j = col; j >= 0 && i < N; i++, j--) 

        if (board[i][j]) 

            return false; 

  

    return true; 

} 

// The function that solves the problem using backtracking 

bool solveNQUtil(int board[N][N], int col) 

{ 

    if (col >= N) 

        return true; 

  

   

    for (int i = 0; i < N; i++) { 

       //if it is safe to place the queen at position i,col -> place it

        if (isSafe(board, i, col)) { 

         

            board[i][col] = 1; 

  

         

            if (solveNQUtil(board, col + 1)) 

                return true; 

  //backtrack if the above condition is false

            board[i][col] = 0; // BACKTRACK 

        } 

    } 

  

   

    return false; 

} 

// driver program to test above function 

int main() 

{ 

     int board[N][N] = { { 0, 0, 0, 0 }, 

                        { 0, 0, 0, 0 }, 

                        { 0, 0, 0, 0 }, 

                        { 0, 0, 0, 0 } }; 

  

    if (solveNQUtil(board, 0) == false) { 

        printf("Solution does not exist"); 

        return 0; 

    } 

  

    printSolution(board); 

    return true; 

    return 0; 

}

File: Java/nQueenProblem.java
import java.util.Arrays;
 
class Main
{
    // Function to check if two queens threaten each other or not
    private static boolean isSafe(char[][] mat, int r, int c)
    {
        // return false if two queens share the same column
        for (int i = 0; i < r; i++)
        {
            if (mat[i][c] == 'Q') {
                return false;
            }
        }
 
        // return false if two queens share the same `` diagonal
        for (int i = r, j = c; i >= 0 && j >= 0; i--, j--)
        {
            if (mat[i][j] == 'Q') {
                return false;
            }
        }
 
        // return false if two queens share the same `/` diagonal
        for (int i = r, j = c; i >= 0 && j < mat.length; i--, j++)
        {
            if (mat[i][j] == 'Q') {
                return false;
            }
        }
 
        return true;
    }
 
    private static void printSolution(char[][] mat)
    {
        for (char[] chars: mat) {
            System.out.println(Arrays.toString(chars).replaceAll(",", ""));
        }
        System.out.println();
    }
 
    private static void nQueen(char[][] mat, int r)
    {
        // if `N` queens are placed successfully, print the solution
        if (r == mat.length)
        {
            printSolution(mat);
            return;
        }
 
        // place queen at every square in the current row `r`
        // and recur for each valid movement
        for (int i = 0; i < mat.length; i++)
        {
            // if no two queens threaten each other
            if (isSafe(mat, r, i))
            {
                // place queen on the current square
                mat[r][i] = 'Q';
 
                // recur for the next row
                nQueen(mat, r + 1);
 
                // backtrack and remove the queen from the current square
                mat[r][i] = '–';
            }
        }
    }
 
    public static void main(String[] args)
    {
        // `N × N` chessboard
        int N = 8;
 
        // `mat[][]` keeps track of the position of queens in
        // the current configuration
        char[][] mat = new char[N][N];
 
        // initialize `mat[][]` by `-`
        for (int i = 0; i < N; i++) {
            Arrays.fill(mat[i], '–');
        }
 
        nQueen(mat, 0);
    }
}
